### PR TITLE
Add data source gitlab_groups

### PR DESCRIPTION
**Pull request by l-with**
## Description

This PR solves issue #1257 .

I added a data source gitlab_groups equivalent to gitlab_users. This is useful to retrieve all groups of an gitlab instance.

### PR Checklist Acknowledgement

- [x] I acknowledge that all of the following items are true, where applicable:
  - Resource attributes match 1:1 the names and structure of the API resource in [the GitLab API documentation](https://docs.gitlab.com/ee/api/).
  - [Examples](https://github.com/gitlabhq/terraform-provider-gitlab/tree/main/examples) are updated with:
    - A \*.tf file for the resource/s with at least one usage example
    - A \*.sh file for the resource/s with an import example (if applicable)
  - New resources have at minimum a basic test with three steps:
    - Create the resource
    - Update the attributes
    - Import the resource
  - No new `//lintignore` comments were copied from existing code. (Linter rules are meant to be enforced on new code.)

### CODE REVIEW

**Comment by l-with** [comment]
@timofurrer Thank you very much for your attention! Please be so kind to approve the waiting workflows.

**Comment by l-with** [comment]
The PR was added to milestone v3.19.0. I received no comments.
The PR passed all checks.
Please tell me, if there is anything I can do.

File: internal/provider/data_source_gitlab_groups_test.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+func TestAccDataSourceGitlabGroups_basic(t *testing.T) {
+	rInt := acctest.RandInt()
+	rInt2 := acctest.RandInt()
+
+	resource.ParallelTest(t, resource.TestCase{
+		ProviderFactories: providerFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceGitlabGroupsConfig(rInt, rInt2),
+				Check: resource.ComposeTestCheckFunc(
```
Comment (armsnyder) [inline internal/provider/data_source_gitlab_groups_test.go:23]: For our provider tests, we perform setup prior to calling `resource.ParallelTest()`, to isolate the resource being tested. Example: https://github.com/gitlabhq/terraform-provider-gitlab/blob/2f06bac4d85ce66bbffdd6b9dc60d319782166be/internal/provider/resource_gitlab_project_approval_rule_test.go#L30

Could you make this change please? This should drastically reduce the number of steps and amount of TF config you need.

**Review by armsnyder — CHANGES_REQUESTED** [review]
Thanks for contributing! The code looks good to me. Just one note on the tests.

File: internal/provider/data_source_gitlab_groups_test.go (hunk 1 above)
Comment (l-with) [reply internal/provider/data_source_gitlab_groups_test.go]: Thank you for the suggestion. 
I need to create groups with specific names for testing the search of the data_source. 
With groups created by testAccCreateGroups, it is only possible to search for "acctest-group". The result contains also groups from parallel running tests and thus the number of results can not be verified.
To fix this I implemented a helper function testAccCreateGroupsWithPrefix and refactored the test with this function.

I hope that matches your intention.

File: internal/provider/data_source_gitlab_groups_test.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	testAccCreateGroupsWithPrefix(t, 42, prefixLotsOf)
+
+	resource.ParallelTest(t, resource.TestCase{
+		ProviderFactories: providerFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceGitlabGroupsConfigSearchSort(prefixFoo),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.gitlab_groups.foos", "groups.#", "2"),
+					resource.TestCheckResourceAttr("data.gitlab_groups.foos", "groups.0.name", groupsFoo[0].Name),
```
Comment (armsnyder) [inline internal/provider/data_source_gitlab_groups_test.go:27]: Could you validate the group attributes here as well?

**Review by armsnyder — CHANGES_REQUESTED** [review]

File: internal/provider/data_source_gitlab_groups_test.go (hunk 1 above)
Comment (armsnyder) [reply internal/provider/data_source_gitlab_groups_test.go]: Thanks

File: internal/provider/data_source_gitlab_groups_test.go (hunk 2 above)
Comment (l-with) [reply internal/provider/data_source_gitlab_groups_test.go:30]: I committed validating the group attributes.

Comment (armsnyder) [reply internal/provider/data_source_gitlab_groups_test.go:30]: Thanks, it looks good. 👍 Could you make sure the CI workflow is passing? There is a guide in the CONTRIBUTING.md, but tl;dr you can run `make reviewable` from the console.

Comment (l-with) [reply internal/provider/data_source_gitlab_groups_test.go:30]: I committed the change resulting from "make reviewable".
I also run "make testacc", but cannot run "make testacc SERVICE=gitlab-ee".

**Review by armsnyder — DISMISSED** [review]

**Review by timofurrer — DISMISSED** [review]
LGTM 🚢

**Review by timofurrer — APPROVED** [review]

**Pull request closed without merging**